src/group_cmd.cpp

```diff
@@ -496,6 +502,192 @@ static void AddVehicleToGroup(Vehicle *v, GroupID new_g)
+			if (std::find_if(unique_orders.begin(), unique_orders.end(), [order](Order *o) { return o->GetDestination() == order->GetDestination(); }) == unique_orders.end()) {
+				unique_orders.push_back(order);
+			}
+		}
+	}
+
+	if (unique_orders.empty()) return_cmd_error(STR_ERROR_GROUP_CAN_T_CREATE_NAME);
+
+	/* Create the name */
+
+	static char str[71] = { "" };  // 5 + 31 + 3 + 31 + 1: ["cargo abbreviation"] "town/station name max. length" - "town/station name max. length""\0"
+
+	StringID cargo_abbreviation = INVALID_STRING_ID;
+	StringID cargo_name = INVALID_STRING_ID;
+
+	for (Vehicle *u = v; u != nullptr; u = u->Next()) {
+		if (u->cargo_cap == 0) continue;
+
+		const CargoSpec *cs = CargoSpec::Get(u->cargo_type);
+		cargo_abbreviation = cs->abbrev;
+		cargo_name = cs->name;
+		break;
+	}
+
+	if (cargo_abbreviation == INVALID_STRING_ID || cargo_name == INVALID_STRING_ID) return_cmd_error(STR_ERROR_GROUP_CAN_T_CREATE_NAME);
+
+	if (_settings_client.gui.specific_group_name == 1) { // Use station names
+
+		static char stationname_first[MAX_LENGTH_STATION_NAME_CHARS] = { "" };
+		static char stationname_last[MAX_LENGTH_STATION_NAME_CHARS] = { "" };
+
+		SetDParam(0, unique_orders.front()->GetDestination());
+		GetString(stationname_first, STR_STATION_NAME, lastof(stationname_first));
+
+		SetDParam(0, unique_orders.back()->GetDestination());
+		GetString(stationname_last, STR_STATION_NAME, lastof(stationname_last));
+
+		SetDParam(0, cargo_abbreviation);
+		SetDParam(1, unique_orders.front()->GetDestination());
+		SetDParam(2, unique_orders.back()->GetDestination());
+		GetString(str, STR_GROUP_SPECIFIC_NAME_STATION, lastof(str));
+
+	}
+	else { //Use town names
+
+		Station *station_first = Station::GetIfValid(unique_orders.front()->GetDestination());
+		Station *station_last = Station::GetIfValid(unique_orders.back()->GetDestination());
+
+		if (station_last == nullptr || station_first == nullptr) return_cmd_error(STR_ERROR_GROUP_CAN_T_CREATE_NAME);
+
+		Town *town_first = station_first->town;
+		Town *town_last = station_last->town;
+
+		if (town_first->index == town_last->index) { // First and last station belong to the same town
+			SetDParam(0, cargo_abbreviation);
+			SetDParam(1, town_first->index);
+			GetString(str, STR_GROUP_SPECIFIC_NAME_TOWN_LOCAL, lastof(str));
+		}
+		else {
+			static char townname_first[MAX_LENGTH_TOWN_NAME_CHARS] = { "" };
+			static char townname_last[MAX_LENGTH_TOWN_NAME_CHARS] = { "" };
+
+			SetDParam(0, town_first->index);
+			GetString(townname_first, STR_TOWN_NAME, lastof(townname_first));
+
+			SetDParam(0, town_last->index);
+			GetString(townname_last, STR_TOWN_NAME, lastof(townname_last));
+
+			SetDParam(0, cargo_abbreviation);
+			SetDParam(1, town_first->index);
+			SetDParam(2, town_last->index );
+			GetString(str, STR_GROUP_SPECIFIC_NAME_TOWN, lastof(str));
+		}
+	}
+
+	// Remove the 'tiny font' formatting
+	std::string s(str);
+	s = s.erase(1, 3);
+	strecpy(&str[0], s.c_str(), lastof(str));
```

Comment:
This is dirty. There *must* be a better way...

Reply:
Maybe do you have any suggestions? :)

Reply:
I changed it. I hope the new approach is better.